feat: Implement Locations endpoint for fetching active EVSEs

satumahayato010 opened the pull request:
## Message
お世話になっております。チャレンジ課題に取り組ませていただきましたので、課題の提出をさせていただきます。
REPORT.mdに、簡易な実装の説明を記載しておりますが、さらなる詳細や実装の意図などのご質問ございましたら、遠慮なくお知らせくださいませ。

お忙しい中大変恐縮ですが、お手隙の際にご確認よろしくお願いいたします。

Commit:
feat: Implement Locations endpoint for fetching active EVSEs

RyuyaIshibashi commented:
@satumahayato010
チャレンジのご対応をいただき、ありがとうございます。
以下のコメントにご留意くださいませ。

https://github.com/enechange/go-challenge/blob/161cc9b5b66a2f93e40636bbf8a227a05043cabc/README.md?plain=1#L12

satumahayato010 commented:
@RyuyaIshibashi
REPORT.mdへの記述指定を見落としてしまっていたようです、すみません！確認不足でした…
記述しきれなかった部分もあるため、不足も含め、REPORT.mdに追記し、更新したいと思います！

rykhr commented:
アーキテクチャにあまり詳しくないので単純な疑問なのですが、
クエリサービスに関して抽象と実装を以下のように分けている理由は何でしょうか？
- 抽象 -> internal/application/query/active_evse_location_query.go
- 実装 -> internal/infrastructure/query/active_evse_location_query_gorm.go

rykhr commented on internal/application/query/active_evse_location_query_test.go:
```diff
@@ -0,0 +1,95 @@
+			m := mock.NewMockActiveEVSELocationQueryService(ctrl)
+			tt.prepareMockFn(m)
+
+			locations, err := m.FindLocationsWithActiveEVSE(context.Background(), tt.latitude, tt.longitude, tt.radius)
```
インターフェースのメソッドをmockしているため、[FindLocationsWithActiveEVSEの具体的な実装](https://github.com/enechange/go-challenge/pull/4/files#diff-f35a4dad707293f0edbd1666c8f548c00ccb60a5ce400b057f192df33c08ce7dR21-R52)の中をテスト時に通っておらず、テストとしての効果が薄いように感じます。

internal/infrastructure/query/active_evse_location_query_gorm.go に対応するテストファイルを作成し、具体的な実装をテストするのはいかがでしょうか。

satumahayato010 replied:
おっしゃる通りだと思いました。現在のテストではモックを使用してインターフェースの振る舞いのみを確認しており、実際のデータアクセスのコードのカバレッジが不足してました…

ご指摘の通り、具体的な実装のテストを追加したいと思います！

satumahayato010 replied:
internal/infrastructure/query/active_evse_location_query_gorm.go に対応するテストを作成しましたので、お手隙の際にご確認よろしくお願いします！

satumahayato010 commented:
@kihara-ori 
抽象層と実装層を分けている理由は、将来的な変更に柔軟に対応できるようにするためです！

抽象層にインターフェースを定義することで、アプリケーションのロジックが特定のDBやクエリに依存することなく、さまざまなDBやクエリで機能できるようになります！

例えば、現在はGormを利用していますが、DB技術の進化やPJの要求が変化し、NoSQLや異なるORMに切り替える必要がある場合でも、アプリケーション層を修正することなく、実装層のみを変更すればよくなるため、変更にかかるコストとリスクを最小限に抑えることができるなどのメリットがあるかと思っています！

Commit:
refactor: Update constructor to accept DB instance and adjust router setup

Commit:
add: test to activeEvseLocationQueryGorm

Commit:
refactor: Move AvailableEVSELocation to DTO and extract mapping logic

Moved AvailableEVSELocation struct from domain to application layer DTO and extracted data mapping to a new mapper file for better separation of concerns

Commit:
refactor: Update Status field type in AvailableEVSELocation DTO

Changed the Status field in the AvailableEVSELocation struct from an int to the domain.Status type to align with domain modeling and ensure type safety

Commit:
docs: Update REPORT.md with detailed architecture explanation

satumahayato010 commented:
@RyuyaIshibashi @kihara-ori @ito-ori 
石橋さん、技術的な決定などをREPORT.mdに記述するように修正いたしました！

お三方、お忙しい中レビューしていただきありがとうございました🙇
新たに修正した内容コミットしていますので、お手隙の際にでもご確認よろしくお願いいたします！

RyuyaIshibashi commented on server.go:
```diff
@@ -13,7 +14,9 @@ import (
-	cnf := config.GetConfig()
+	cnf := configs.GetConfig()
+
+	validators.RegisterCustomValidations()
```
将来別のエンドポイント実装をする際、`latitude`, `longitude`カラムに別のバリーデーションを設定することは可能でしょうか？
